### PR TITLE
Feature/neural net tuning

### PR DESCRIPTION
**Pull request by evenmm**
New things in this commit: 
- Updated inclusion criteria to include filtering on low water levels and long eq_times
- The NMR lipids databank has changed the way they do initialization of the databank
- Added simulated noise to input (form factors) 
- Added comparison with convolutional network 
- New build functions for both network types in new_functions.py
- Hyperparameter tuning for both network types
- Plotting mean and median absolute deviation in test set

Want input on: 
- The way hyperparameter tuning is set up 
- Additional parameters to tune over
- The added functions in new_functions.py: Generalizability, clarity, and best practices
- If anything else could be moved to new_functions.py
- Evaluation metrics or plots you would like to see

Of interest later but of scope on the current contract: 
- Reintroducing callbacks in the cross validation setup: for early stopping, plotting of errors and prediciton visualization during training for interpretability

### CODE REVIEW

File: new_functions.py [hunk 1]
```diff
@@ -135,54 +184,88 @@ def extrapolate_Y(y_vector: np.ndarray, length_of_padded_data: int) -> np.ndarra
...
+    """
+    Rescale values to [-0.5, 0.5] range consistent with provided global min and max values
+
+    :param values: values to rescale
+    :param min_value: minimum of original range
+    :param max_value: maximum of original range
+
+    :return: Values rescaled to range of [-0.5, 0.5]
+    """
+    return (values - min_value) / (max_value - min_value) - 0.5
```
Comment (borgesa) [inline new_functions.py:199]:
```suggestion
    if min_value == max_value:
        raise ValueError("min_value and max_value must be different to avoid division by zero.")

    return (values - min_value) / (max_value - min_value) - 0.5
```

.. that said, not yet handling exceptions in the caller of this, but... Might as well prepare for it.

File: new_functions.py [hunk 2]
```diff
@@ -135,54 +184,88 @@ def extrapolate_Y(y_vector: np.ndarray, length_of_padded_data: int) -> np.ndarra
...
+
+    rescaled_start = rescale_to_zero_centered_unit_range(
+        uniform_x_range[0], global_min_td_x, global_max_td_x
+    )
+    rescaled_end = rescale_to_zero_centered_unit_range(
+        uniform_x_range[-1], global_min_td_x, global_max_td_x
+    )
+    rescaled_uniform_x_range = np.linspace(
+        rescaled_start, rescaled_end, len(uniform_x_range)
+    ).reshape(-1, 1)
```
Comment (borgesa) [inline new_functions.py:254]: Are the input dimensions the same across all?

If so:
Maybe rewrite this, vectorized? This would avoid the loops and overall be cleaner to read.

File: new_functions.py [hunk 3]
```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
...
+
+def add_noise_to_form_factor(
+    form_factor_df: pd.DataFrame,
+    steepness: float = 10.0,
+    max_noise_magnitude: float = 1.0,
+    probability_of_noise: float = 1.0,
+) -> pd.DataFrame:
+    """
+    Adds random noise to each row of the input DataFrame, with noise amplitude determined by a pure sigmoid function (scaled by `max_noise_magnitude`) across the columns. The amplitude starts near 0 in the first column and approaches `max_noise_magnitude` in the last column.
+
```
Comment (borgesa) [inline new_functions.py:334]: Is this emulating the noise in the original samples?

If so, maybe mention this in the docstring. Something like: 
"The added noise emulates the nature of the noise seen in the measured data."

File: new_functions.py [hunk 4]
```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
...
+    num_rows, num_cols = form_factor_df.shape
+
+    # Array of values from 0 to 1 in the dimension of q, to make the sigmoid
+    fraction_array = np.linspace(0.0, 1.0, num_cols)
+
+    # Noise amplitudes follow the logistic (sigmoid) function, scaled by max_noise_magnitude
+    # At fraction=0, amplitude ~ 0; at fraction=1, amplitude ~ max_noise_magnitude
+    noise_amplitudes = max_noise_magnitude * (
+        1.0 / (1.0 + np.exp(-steepness * (fraction_array - 0.5)))
+    )
```
Comment (borgesa) [inline new_functions.py:355]:
```suggestion
    # Noise amplitudes follow the logistic (sigmoid) function, scaled by max_noise_magnitude
    # At fraction=0, amplitude ~ 0; at fraction=1, amplitude ~ max_noise_magnitude
    fraction_array = np.linspace(0.0, 1.0, num_cols)
    noise_amplitudes = max_noise_magnitude * (
        1.0 / (1.0 + np.exp(-steepness * (fraction_array - 0.5)))
    )
```
(just a bit more compact. the linspace call doesn't need explanation)

File: new_functions.py [hunk 5]
```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
...
+
+    # Generate a mask indicating which rows should receive noise (80% probability)
+    noise_mask = np.random.rand(num_rows) < probability_of_noise
+
+    # Generate random noise with per-column standard deviation according to noise_amplitudes
+    noise = np.random.normal(loc=0.0, scale=scale_array)
+
+    # Add noise to the original data
+    noisy_values = form_factor_df.values.copy()
+    noisy_values[noise_mask] += noise[noise_mask]
```
Comment (borgesa) [inline new_functions.py:368]:
```suggestion
# Generate a mask indicating which rows should receive noise (80% probability)
noise_mask = np.random.rand(num_rows) < probability_of_noise

# Generate random noise with per-column standard deviation according to noise_amplitudes
noise = np.random.normal(loc=0.0, scale=noise_amplitudes, size=(num_rows, num_cols))

# Apply noise only to the selected rows
noisy_values = form_factor_df.values.copy()
noisy_values = np.where(noise_mask[:, np.newaxis], noisy_values + noise, noisy_values)
```

File: new_functions.py [hunk 6]
```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
...
+            activation=activation_function,
+        )(x)
+        x = layers.MaxPooling1D(pool_size=2)(x)
+
+    # Flatten and final dense layer
+    x = layers.Flatten()(x)
+    outputs = layers.Dense(output_dim, activation=None)(x)
+
+    model = models.Model(inputs=keras_input, outputs=outputs)
+    model.compile(optimizer="adam", loss="mse", metrics=["mae"])
```
Comment (borgesa) [inline new_functions.py:444]: Why not use `adamw` also here?

Ref from FC model:
`optimizer=tf.keras.optimizers.AdamW(amsgrad=True),`

**Review by borgesa — COMMENTED** [review]
I started on the wrong for this review...

Still need to add comments to the notebook, but releasing the comments this far, mainly on `new_functions.py` and the README.

File: FormFactor_to_TotalDensity_model_training_walkthrough.ipynb
Comment (borgesa) [inline FormFactor_to_TotalDensity_model_training_walkthrough.ipynb:1]: I heard you discussed the notebook with Omar a bit already.
This could be split into more bite-size notebooks, separating different tasks (as you already discussed).

File: new_functions.py [hunk 7]
```diff
@@ -194,7 +277,169 @@ def plot_training_trajectory(ax: Axes, history: object) -> Axes:
...
+    """
+    keras_input = keras.Input(shape=(input_dim,))
+    x = keras_input
+
+    # Create as many hidden layers as specified in hidden_layer_dims
+    for dim in hidden_layer_dims:
+        x = layers.Dense(dim, activation="relu")(x)
+
+    # Final output layer
+    output_layer = layers.Dense(output_dim, activation="linear")(x)
```
Comment (mhwaage) [inline new_functions.py:398]:
```python
from functools import reduce

hidden_layers = reduce(lambda dim, x: layers.Dense(dim, activation="relu"), hidden_layer_dims, keras.Input(shape=(input_dim,)))

output_layer = layers.Dense(output_dim, activation="linear")(hidden_layers)
```
should be equivalent

File: pyproject.toml [hunk 8]
```diff
@@ -13,8 +14,11 @@ matplotlib = "^3.9.2"
 tqdm = "^4.67.0"
 pyyaml = "^6.0.2"
 mdanalysis = "^2.7.0"
-scikit-learn = "^1.5.2"
+scikit-learn = ">=1.5.2,<1.5.3"
```
Comment (mhwaage) [inline pyproject.toml:17]: isn't this just 1.5.2?
```suggestion
scikit-learn = "1.5.2"
```

**Review by mhwaage — COMMENTED** [review]
only adding comments, don't know enough about the field to do anything beyond that

**Review by evenmm — COMMENTED** [review]
Most suggested changes have been addressed locally and will be applied in the next commit. One change in pyproject.toml is accepted directly.
